File: android/airship-framework-proxy/src/test/java/com/urbanairship/android/framework/proxy/AttributeOperationTest.kt

package com.urbanairship.android.framework.proxy

import com.urbanairship.channel.AttributeEditor
import com.urbanairship.json.JsonValue
import com.urbanairship.json.jsonMapOf
import io.mockk.confirmVerified
import io.mockk.mockk
import io.mockk.verify
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.util.*

@RunWith(RobolectricTestRunner::class)
public class AttributeOperationTest {

    @Test
    public fun testRemove() {
        val json = """
            {
                "key": "some attribute",
                "action": "remove"
            }
        """

        val parsed = AttributeOperation(json = JsonValue.parseString(json).requireMap())

        val expected = AttributeOperation(
            attribute = "some attribute",
            value = null,
            action = AttributeOperationAction.REMOVE,
            valueType = null
        )

        assertEquals(expected, parsed)
    }

    @Test
    public fun testRemoveJson() {
        val json = """
            {
                "key": "some attribute",
                "action": "remove",
                "instance_id": "Some instance"
            }
        """

        val parsed = AttributeOperation(json = JsonValue.parseString(json).requireMap())

        val expected = AttributeOperation(
            attribute = "some attribute",
            value = null,
            action = AttributeOperationAction.REMOVE,
            valueType = null,
            instanceId = "Some instance"
        )

        assertEquals(expected, parsed)
    }

    @Test
    public fun testJson() {
        val json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": { "foo": "bar" },
                "type": "json",
                "instance_id": "instance"
            }
        """

        val parsed = AttributeOperation(json = JsonValue.parseString(json).requireMap())

        val expected = AttributeOperation(
            attribute = "some attribute",
            value = jsonMapOf("foo" to "bar").toJsonValue(),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.JSON,
            instanceId = "instance"
        )

        assertEquals(expected, parsed)
    }

    @Test
    public fun testDate() {
        val json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": 1682681877000,
                "type": "date"
            }
        """

        val parsed = AttributeOperation(json = JsonValue.parseString(json).requireMap())

        val expected = AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt(1682681877000),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.DATE
        )

        assertEquals(expected, parsed)
    }

    @Test
    public fun testString() {
        val json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": "neat",
                "type": "string"
            }
        """

        val parsed = AttributeOperation(json = JsonValue.parseString(json).requireMap())

        val expected = AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt("neat"),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.STRING
        )

        assertEquals(expected, parsed)
    }

    @Test
    public fun testNumber() {
        val json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": 10.4,
                "type": "number"
            }
        """

        val parsed = AttributeOperation(json = JsonValue.parseString(json).requireMap())

        val expected = AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt(10.4),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.NUMBER
        )

        assertEquals(expected, parsed)
    }

    @Test
    public fun testApplyJson() {
        val editor = mockk<AttributeEditor>(relaxed = true)

        val operation  = AttributeOperation(
            attribute = "some attribute",
            value = jsonMapOf("foo" to "bar").toJsonValue(),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.JSON,
            instanceId = "instance id",
            expiry = Date(1000)
        )

        operation.applyOperation(editor)

        verify { editor.setAttribute(
            attribute = "some attribute",
            instanceId = "instance id",
            expiration = Date(1000),
            json = jsonMapOf("foo" to "bar")
        ) }
        confirmVerified(editor)
    }

    @Test
    public fun testApplyString() {
        val editor = mockk<AttributeEditor>(relaxed = true)

        val operation  = AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt("neat"),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.STRING
        )

        operation.applyOperation(editor)

        verify { editor.setAttribute("some attribute", "neat") }
        confirmVerified(editor)
    }

    @Test
    public fun testApplyNumber() {
        val editor = mockk<AttributeEditor>(relaxed = true)

        val operation  = AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt(100.1),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.NUMBER
        )

        operation.applyOperation(editor)

        verify { editor.setAttribute("some attribute", 100.1) }
        confirmVerified(editor)
    }

    @Test
    public fun testApplyDate() {
        val editor = mockk<AttributeEditor>(relaxed = true)

        val operation  = AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt(1682681877000),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.DATE
        )

        operation.applyOperation(editor)

        verify { editor.setAttribute("some attribute", Date(1682681877000)) }
        confirmVerified(editor)
    }

    @Test
    public fun testApplyRemove() {
        val editor = mockk<AttributeEditor>(relaxed = true)

        val operation = AttributeOperation(
            attribute = "some attribute",
            value = null,
            action = AttributeOperationAction.REMOVE,
            valueType = null
        )

        operation.applyOperation(editor)

        verify { editor.removeAttribute("some attribute") }
        confirmVerified(editor)
    }

    @Test
    public fun testApplyRemoveJson() {
        val editor = mockk<AttributeEditor>(relaxed = true)

        val operation = AttributeOperation(
            attribute = "some attribute",
            value = null,
            action = AttributeOperationAction.REMOVE,
            valueType = null,
            instanceId = "some instance"
        )

        operation.applyOperation(editor)

        verify { editor.removeAttribute("some attribute", "some instance") }
        confirmVerified(editor)
    }

    @Test(expected = java.lang.Exception::class)
    public fun testApplyInvalidString() {
        val editor = mockk<AttributeEditor>()

        AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt(1682681877000),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.STRING
        ).applyOperation(editor)
    }


    @Test(expected = java.lang.Exception::class)
    public fun testApplyNullString() {
        val editor = mockk<AttributeEditor>()

        AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.NULL,
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.STRING
        ).applyOperation(editor)
    }

    @Test(expected = java.lang.Exception::class)
    public fun testApplyInvalidNumber() {
        val editor = mockk<AttributeEditor>()

        AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt("nope"),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.NUMBER
        ).applyOperation(editor)
    }


    @Test(expected = java.lang.Exception::class)
    public fun testApplyNullNumber() {
        val editor = mockk<AttributeEditor>()

        AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.NULL,
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.NUMBER
        ).applyOperation(editor)
    }

    @Test(expected = java.lang.Exception::class)
    public fun testApplyInvalidDate() {
        val editor = mockk<AttributeEditor>()

        AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.wrapOpt("nope"),
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.DATE
        ).applyOperation(editor)
    }


    @Test(expected = java.lang.Exception::class)
    public fun testApplyNullDate() {
        val editor = mockk<AttributeEditor>()

        AttributeOperation(
            attribute = "some attribute",
            value = JsonValue.NULL,
            action = AttributeOperationAction.SET,
            valueType = AttributeValueType.DATE
        ).applyOperation(editor)
    }
}

File: ios/AirshipFrameworkProxyTests/AttributeOperationTest.swift
import Testing

@testable import AirshipFrameworkProxy
import AirshipKit

struct AttributeOperationTest {
    private let editor = TestEditor()

    @Test
    func testRemove() throws {
        let json = """
            {
                "key": "some attribute",
                "action": "remove"
            }
        """


        let parsed = try JSONDecoder().decode(AttributeOperation.self, from: json.data(using: .utf8)!)

        let expected = AttributeOperation(
            action: .removeAttribute,
            attribute: "some attribute",
            value: nil,
            valueType: nil
        )

        assert(expected == parsed)
    }

    @Test
    func testRemoveJSON() throws {
        let json = """
            {
                "key": "some attribute",
                "instance_id": "some instance id",
                "action": "remove"
            }
        """

        let parsed = try JSONDecoder().decode(AttributeOperation.self, from: json.data(using: .utf8)!)

        let expected = AttributeOperation(
            action: .removeAttribute,
            attribute: "some attribute",
            value: nil,
            valueType: nil,
            instanceID: "some instance id"
        )

        assert(expected == parsed)
    }

    @Test
    func tesSetDate() throws {
        let json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": 1682681877000,
                "type": "date"
            }
        """

        let parsed = try JSONDecoder().decode(AttributeOperation.self, from: json.data(using: .utf8)!)

        let expected = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: try AirshipJSON.wrap(1682681877000),
            valueType: .date
        )

        assert(expected == parsed)
    }

    @Test
    func testSetString() throws {
        let json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": "neat",
                "type": "string"
            }
        """


        let parsed = try JSONDecoder().decode(AttributeOperation.self, from: json.data(using: .utf8)!)

        let expected = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: AirshipJSON.string("neat"),
            valueType: .string
        )

        assert(expected == parsed)
    }

    @Test
    func testSetNumber() throws {
        let json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": 10.4,
                "type": "number"
            }
        """


        let parsed = try JSONDecoder().decode(AttributeOperation.self, from: json.data(using: .utf8)!)

        let expected = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: AirshipJSON.number(10.4),
            valueType: .number
        )

        assert(expected == parsed)
    }

    @Test
    func testSetJSON() throws {
        let json = """
            {
                "key": "some attribute",
                "action": "set",
                "value": {
                    "foo": "bar"
                },
                "type": "json",
                "instance_id": "some instance id"
            }
        """


        let parsed = try JSONDecoder().decode(AttributeOperation.self, from: json.data(using: .utf8)!)

        let expected = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: AirshipJSON.object(["foo": .string("bar")]),
            valueType: .json,
            instanceID: "some instance id"
        )

        assert(expected == parsed)
    }

    @Test
    func testApplyRemove() async throws {
        let operation = AttributeOperation(
            action: .removeAttribute,
            attribute: "some attribute",
            value: nil,
            valueType: nil
        )

        try await confirmation { continuation in
            editor.onMutation = { mutation in
                assert(mutation == .remove(attribute: "some attribute"))
                continuation.confirm()
            }

            try operation.apply(editor: editor)
        }
    }

    @Test
    func testApplyRemoveJSON() async throws {
        let operation = AttributeOperation(
            action: .removeAttribute,
            attribute: "some attribute",
            value: nil,
            valueType: nil,
            instanceID: "some instance"
        )

        try await confirmation { continuation in
            editor.onMutation = { mutation in
                assert(mutation == .remove(attribute: "some attribute", instanceID: "some instance"))
                continuation.confirm()
            }

            try operation.apply(editor: editor)
        }
    }

    @Test
    func testApplyJSON() async throws {
        let operation = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: .object(["neat": .string("story")]),
            valueType: .json,
            instanceID: "some instance",
            expirationMilliseconds: 1000
        )

        try await confirmation { continuation in
            editor.onMutation = { mutation in
                let expected = TestEditor.Mutation.set(
                    value: ["neat": AirshipJSON.string("story")],
                    attribute: "some attribute",
                    instanceID: "some instance",
                    expiration: Date(timeIntervalSince1970: 1.0)
                )
                assert(mutation == expected)
                continuation.confirm()
            }

            try operation.apply(editor: editor)
        }
    }


    @Test
    func testApplySetString() async throws {
        let operation = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: .string("neat"),
            valueType: .string
        )

        try await confirmation { continuation in
            editor.onMutation = { mutation in
                let expected = TestEditor.Mutation.set(
                    value: "neat",
                    attribute: "some attribute"
                )
                assert(mutation == expected)
                continuation.confirm()
            }

            try operation.apply(editor: editor)
        }
    }

    @Test
    func testApplyNumber() async throws {
        let operation = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: .number(100.1),
            valueType: .number
        )

        try await confirmation { continuation in
            editor.onMutation = { mutation in
                let expected = TestEditor.Mutation.set(
                    value: 100.1,
                    attribute: "some attribute"
                )
                assert(mutation == expected)
                continuation.confirm()
            }

            try operation.apply(editor: editor)
        }
    }

    @Test
    func testApplyDate() async throws {
        let operation = AttributeOperation(
            action: .setAttribute,
            attribute: "some attribute",
            value: .number(1682681877000),
            valueType: .date
        )

        try await confirmation { continuation in
            editor.onMutation = { mutation in
                let expected = TestEditor.Mutation.set(
                    value: Date(timeIntervalSince1970: 1682681877),
                    attribute: "some attribute"
                )
                assert(mutation == expected)
                continuation.confirm()
            }

            try operation.apply(editor: editor)
        }
    }

    @Test
    func testApplyInvalidString() throws {
        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "number string attribute",
                value: .number(10.0),
                valueType: .string
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}

        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "null string attribute",
                value: .null,
                valueType: .string
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}

        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "dictionary string attribute",
                value: try! AirshipJSON.wrap(["cool": "beans"]),
                valueType: .string
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}
    }

    @Test
    func testApplyInvalidNumber() throws {
        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "string number attribute",
                value: .string("hello"),
                valueType: .number
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}

        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "null number attribute",
                value: .null,
                valueType: .number
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}


        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "dictionary number attribute",
                value: try! AirshipJSON.wrap(["cool": "beans"]),
                valueType: .number
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}

    }

    @Test
    func testApplyInvalidDate() throws {
        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "string date attribute",
                value: .string("1682681877000"),
                valueType: .date
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}

        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "null date attribute",
                value: .null,
                valueType: .date
            ).apply(editor: editor)
            assertionFailure("should throw")
        } catch {}


        do {
            try AttributeOperation(
                action: .setAttribute,
                attribute: "dictionary date attribute",
                value: try! AirshipJSON.wrap(["cool": "beans"]),
                valueType: .date
            ).apply(editor: editor)
        } catch {}
    }
}

fileprivate class TestEditor: AttributeOperationEditor {
    enum Mutation: Equatable, Hashable {
        case set(value: AnyHashable, attribute: String, instanceID: String? = nil, expiration: Date? = nil)
        case remove(attribute: String, instanceID: String? = nil)
    }

    var onMutation: ((Mutation) -> Void)? = nil

    func remove(_ attribute: String) {
        onMutation!(.remove(attribute: attribute))
    }

    func remove(attribute: String, instanceID: String) throws {
        onMutation!(.remove(attribute: attribute, instanceID: instanceID))
    }

    func set(date: Date, attribute: String) {
        onMutation!(.set(value: date, attribute: attribute))
    }

    func set(double: Double, attribute: String) {
        onMutation!(.set(value: double, attribute: attribute))
    }

    func set(string: String, attribute: String) {
        onMutation!(.set(value: string, attribute: attribute))
    }


    func set(json: [String : AirshipKit.AirshipJSON], attribute: String, instanceID: String, expiration: Date?) throws {
        onMutation!(.set(value: json, attribute: attribute, instanceID: instanceID, expiration: expiration))
    }
}
